cni-repair-controller/src/lib.rs

```diff
@@ -0,0 +1,172 @@
+    },
+    Client, Error, Resource, ResourceExt,
+};
+use kubert::Runtime;
+use prometheus_client::{metrics::counter::Counter, registry::Registry};
+use tokio::sync::mpsc::{self, error::TrySendError, Receiver, Sender};
+
+// ERRNO 95: Operation not supported
+pub const UNSUCCESSFUL_EXIT_CODE: i32 = 95;
+
+// If the event channel capacity is reached, the event is dropped, but a new one will be emitted
+// in the pod's next crashloop iteration
+const EVENT_CHANNEL_CAPACITY: usize = 32;
+
+const DATA_PLANE_LABEL: &str = "linkerd.io/control-plane-ns";
+const CONDITION_EVICTED_REASON: &str = "EvictionByEvictionAPI";
+const EVENT_ACTION: &str = "Evicting";
+const EVENT_REASON: &str = "LinkerdCNINotConfigured";
+
+#[derive(Clone, Debug)]
+pub struct Metrics {
+    queue_overflow: Counter<u64>,
+    evicted_pods: Counter<u64>,
+}
+
+pub fn run(rt: &mut Runtime, node_name: String, controller_pod_name: String, metrics: Metrics) {
+    let pod_evts = rt.watch_all::<Pod>(
+        watcher::Config::default()
+            .labels(DATA_PLANE_LABEL)
+            .fields(&format!("spec.nodeName={node_name}")),
+    );
+    let (tx, rx) = mpsc::channel(EVENT_CHANNEL_CAPACITY);
+    tokio::spawn(process_events(pod_evts, tx, metrics.clone()));
+
+    let client = rt.client();
+    tokio::spawn(process_pods(client, controller_pod_name, rx, metrics));
+}
+
+async fn process_events(
+    pod_evts: impl Stream<Item = watcher::Event<Pod>>,
+    tx: Sender<Pod>,
+    metrics: Metrics,
+) {
+    tokio::pin!(pod_evts);
+    while let Some(evt) = pod_evts.next().await {
+        tracing::trace!(?evt);
+        if let watcher::Event::Applied(pod) = evt {
+            let status = if let Some(ref status) = pod.status {
+                status.clone()
+            } else {
+                tracing::info!("Skipped, no status");
+                continue;
+            };
+
+            let terminated = status
+                .init_container_statuses
+                .unwrap_or_default()
+                .iter()
+                .find(|status| status.name == "linkerd-network-validator")
+                .and_then(|status| status.last_state.as_ref())
+                .and_then(|status| status.terminated.as_ref())
+                .map(|term| term.exit_code == UNSUCCESSFUL_EXIT_CODE)
+                .unwrap_or(false);
+
+            let evicted = status
+                .conditions
+                .as_ref()
+                .and_then(|conds| {
+                    conds.iter().find(|cond| {
+                        cond.reason
+                            .as_ref()
+                            .is_some_and(|reason| reason == CONDITION_EVICTED_REASON)
+                    })
+                })
+                .is_some();
+
+            if terminated && !evicted {
+                let namespace = pod.namespace().unwrap();
+                let name = pod.name_any();
+                // this avoids blocking the event loop
+                match tx.try_send(pod.clone()) {
```

Comment:
Note that this is going to clone all of the pod's spec. It would be substantially more efficient to only send a namespace and name over the channel.

Reply:
We use the pod object down the call stack in `publish_k8s_event` to retrieve its ObjectReference. That happens to have name and namespace, so I'll pass that instead.